codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/serde/XmlParserGenerator.kt

```diff
@@ -575,43 +576,49 @@ open class XmlParserGenerator(
                     writer.write(textExpr)
                     return
                 } else {
                     null
                 }
             }
-            ShapeType.TIMESTAMP -> {
+            target.type == ShapeType.TIMESTAMP -> {
                 val trait = member.getTrait<TimestampFormatTrait>() ?: target.getTrait()
                 val tsFormat = trait?.format ?: defaultTimestampFormat
                 val runtimeEnum = tsFormat.toRuntimeEnum(writer)
                 writer.format("#T(#L)", RuntimeTypes.Serde.parseTimestamp, runtimeEnum)
             }
-            ShapeType.BYTE -> writer.format("#T()", RuntimeTypes.Serde.parseByte)
-            ShapeType.SHORT -> writer.format("#T()", RuntimeTypes.Serde.parseShort)
-            ShapeType.INTEGER -> writer.format("#T()", RuntimeTypes.Serde.parseInt)
-            ShapeType.LONG -> writer.format("#T()", RuntimeTypes.Serde.parseLong)
-            ShapeType.FLOAT -> writer.format("#T()", RuntimeTypes.Serde.parseFloat)
-            ShapeType.DOUBLE -> writer.format("#T()", RuntimeTypes.Serde.parseDouble)
-            ShapeType.BIG_DECIMAL -> writer.format("#T()", RuntimeTypes.Serde.parseBigDecimal)
-            ShapeType.BIG_INTEGER -> writer.format("#T()", RuntimeTypes.Serde.parseBigInteger)
-            ShapeType.ENUM -> {
+            target.type == ShapeType.BYTE -> writer.format("#T()", RuntimeTypes.Serde.parseByte)
+            target.type == ShapeType.SHORT -> writer.format("#T()", RuntimeTypes.Serde.parseShort)
+            target.type == ShapeType.INTEGER -> writer.format("#T()", RuntimeTypes.Serde.parseInt)
+            target.type == ShapeType.LONG -> writer.format("#T()", RuntimeTypes.Serde.parseLong)
+            target.type == ShapeType.FLOAT -> writer.format("#T()", RuntimeTypes.Serde.parseFloat)
+            target.type == ShapeType.DOUBLE -> writer.format("#T()", RuntimeTypes.Serde.parseDouble)
+            target.type == ShapeType.BIG_DECIMAL -> writer.format("#T()", RuntimeTypes.Serde.parseBigDecimal)
+            target.type == ShapeType.BIG_INTEGER -> writer.format("#T()", RuntimeTypes.Serde.parseBigInteger)
+            target.type == ShapeType.ENUM || (target.type == ShapeType.STRING && target.hasTrait<EnumTrait>()) -> {
+                println("Shape ${member.defaultName()} is an enum! textExprIsResult is $textExprIsResult")
                 if (!textExprIsResult) {
                     writer.write("#T.fromValue(#L)", ctx.symbolProvider.toSymbol(target), textExpr)
                     return
+                } else {
+                    writer.format("#T { #T.fromValue(it) } ", RuntimeTypes.Serde.parse, ctx.symbolProvider.toSymbol(target)).also {
+                        println("Shape ${member.defaultName()} is an enum, parseFn is $it")
+                    }
                 }
-                writer.format("#T { #T.fromValue(it) } ", RuntimeTypes.Serde.parse, ctx.symbolProvider.toSymbol(target))
             }
```

Comment:
**Suggestion**: Replace `ShapeType.ENUM || (target.type == ShapeType.STRING && target.hasTrait<EnumTrait>())` with [`target.isStringEnumShape`](https://github.com/smithy-lang/smithy-kotlin/blob/1a656f4100a126a236f092a960ead7ba75cfee24/codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/ShapeExt.kt#L163-L174). Also, move it up above the general string case and then you can simplify that branch to just `target.type == StringType.String`.